File: apps/mobile_ios/test/fuel_plan_test.dart
import 'package:flutter_test/flutter_test.dart';

import '../lib/fuel_plan.dart';

// start -> aid (1h, refill) -> cutoff (0.5h, no services) -> finish (0.5h).
List<FuelLegInput> legs() => const [
      FuelLegInput(projectedElapsedS: 0, legDistM: 0, services: []),
      FuelLegInput(projectedElapsedS: 3600, legDistM: 8000, services: ['water', 'food']),
      FuelLegInput(projectedElapsedS: 5400, legDistM: 4000, services: []),
      FuelLegInput(projectedElapsedS: 7200, legDistM: 4000, services: []),
    ];

FuelPlan plan({double carbs = 60, double fluid = 500, double? heat, double? weightKg}) =>
    buildFuelPlan(legs(),
        carbsPerHourG: carbs, fluidPerHourMl: fluid, heatFactor: heat, weightKg: weightKg);

void main() {
  test('carbs + fluid scale with each leg duration', () {
    final p = plan();
    expect(p.legs[1].carbsG, 60);
    expect(p.legs[1].fluidMl, 500);
    expect(p.legs[2].carbsG, 30);
    expect(p.legs[2].fluidMl, 250);
    expect(p.legs[3].carbsG, 30);
  });

  test('zero-duration leg (start) gets zero carbs + fluid', () {
    final p = plan();
    expect(p.legs[0].carbsG, 0);
    expect(p.legs[0].fluidMl, 0);
  });

  test('heat factor bumps fluid but not carbs', () {
    final base = plan();
    final hot = plan(heat: heatFluidFactor);
    expect(hot.legs[1].carbsG, base.legs[1].carbsG);
    expect(hot.legs[1].fluidMl, base.legs[1].fluidMl * heatFluidFactor);
  });

  test('carryToNextAid sums legs up to and including the next refill', () {
    final p = plan();
    expect(p.legs[0].carryToNextAid!.carbsG, 60);
    expect(p.legs[0].carryToNextAid!.fluidMl, 500);
    expect(p.legs[0].carryToNextAid!.gels, 3);
    expect(p.legs[1].carryToNextAid!.carbsG, 60);
    expect(p.legs[1].carryToNextAid!.fluidMl, 500);
    expect(p.legs[1].carryToNextAid!.gels, 3);
  });

  test('carry is present only on the start + refill checkpoints', () {
    final p = plan();
    expect(p.legs[0].carryToNextAid, isNotNull);
    expect(p.legs[1].carryToNextAid, isNotNull);
    expect(p.legs[2].carryToNextAid, isNull);
    expect(p.legs[3].carryToNextAid, isNull);
  });

  test('gel count is ceil(carbs / gelCarbsG)', () {
    final p = plan(carbs: 70);
    expect(p.legs[0].carryToNextAid!.carbsG, 70);
    expect(p.legs[0].carryToNextAid!.gels, (70 / gelCarbsG).ceil());
  });

  test('totals equal the sum of the legs', () {
    final p = plan();
    final sumCarbs = p.legs.fold<double>(0, (a, l) => a + l.carbsG);
    final sumFluid = p.legs.fold<double>(0, (a, l) => a + l.fluidMl);
    expect(p.totalCarbsG, sumCarbs);
    expect(p.totalFluidMl, sumFluid);
    expect(p.totalCarbsG, 120);
    expect(p.totalFluidMl, 1000);
  });

  test('kcal is estimated only when a bodyweight is supplied', () {
    expect(plan().legs[1].kcal, 0);
    final withWeight = plan(weightKg: 70);
    expect((withWeight.legs[1].kcal - 580.16).abs() < 0.01, isTrue);
  });

  test('non-positive intake rates clamp to zero', () {
    final p = plan(carbs: -10, fluid: -5);
    expect(p.totalCarbsG, 0);
    expect(p.totalFluidMl, 0);
  });
}

File: apps/web/src/lib/routes/fuel_plan.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
	buildFuelPlan,
	type FuelLegInput,
	GEL_CARBS_G,
	HEAT_FLUID_FACTOR
} from './fuel_plan';

// start → aid (1h, refill) → cutoff (0.5h, no services) → finish (0.5h).
function legs(): FuelLegInput[] {
	return [
		{ projectedElapsedS: 0, legDistM: 0, services: [] },
		{ projectedElapsedS: 3600, legDistM: 8000, services: ['water', 'food'] },
		{ projectedElapsedS: 5400, legDistM: 4000, services: [] },
		{ projectedElapsedS: 7200, legDistM: 4000, services: [] }
	];
}

const OPTS = { carbsPerHourG: 60, fluidPerHourMl: 500 };

test('carbs + fluid scale with each leg duration', () => {
	const plan = buildFuelPlan(legs(), OPTS);
	// 1h leg → 60 g / 500 ml; the two 0.5h legs → 30 g / 250 ml each.
	assert.equal(plan.legs[1].carbsG, 60);
	assert.equal(plan.legs[1].fluidMl, 500);
	assert.equal(plan.legs[2].carbsG, 30);
	assert.equal(plan.legs[2].fluidMl, 250);
	assert.equal(plan.legs[3].carbsG, 30);
});

test('zero-duration leg (start) gets zero carbs + fluid', () => {
	const plan = buildFuelPlan(legs(), OPTS);
	assert.equal(plan.legs[0].carbsG, 0);
	assert.equal(plan.legs[0].fluidMl, 0);
});

test('heat factor bumps fluid but not carbs', () => {
	const base = buildFuelPlan(legs(), OPTS);
	const hot = buildFuelPlan(legs(), { ...OPTS, heatFactor: HEAT_FLUID_FACTOR });
	assert.equal(hot.legs[1].carbsG, base.legs[1].carbsG);
	assert.equal(hot.legs[1].fluidMl, base.legs[1].fluidMl * HEAT_FLUID_FACTOR);
});

test('carryToNextAid sums legs up to and including the next refill', () => {
	const plan = buildFuelPlan(legs(), OPTS);
	// Out of the start: only the leg arriving at the aid (the next refill).
	assert.deepEqual(plan.legs[0].carryToNextAid, { carbsG: 60, fluidMl: 500, gels: 3 });
	// Out of the aid: cutoff + finish legs (no further refill) → to the end.
	assert.deepEqual(plan.legs[1].carryToNextAid, { carbsG: 60, fluidMl: 500, gels: 3 });
});

test('carry is present only on the start + refill checkpoints', () => {
	const plan = buildFuelPlan(legs(), OPTS);
	assert.ok(plan.legs[0].carryToNextAid); // start
	assert.ok(plan.legs[1].carryToNextAid); // aid (refill)
	assert.equal(plan.legs[2].carryToNextAid, undefined); // cutoff, no services
	assert.equal(plan.legs[3].carryToNextAid, undefined); // finish
});

test('gel count is ceil(carbs / gelCarbsG)', () => {
	// 70 g/hr over the 1h leg out of the start → 70 g → ceil(70/25) = 3 gels.
	const plan = buildFuelPlan(legs(), { carbsPerHourG: 70, fluidPerHourMl: 500 });
	assert.equal(plan.legs[0].carryToNextAid?.carbsG, 70);
	assert.equal(plan.legs[0].carryToNextAid?.gels, Math.ceil(70 / GEL_CARBS_G));
});

test('totals equal the sum of the legs', () => {
	const plan = buildFuelPlan(legs(), OPTS);
	const sumCarbs = plan.legs.reduce((a, l) => a + l.carbsG, 0);
	const sumFluid = plan.legs.reduce((a, l) => a + l.fluidMl, 0);
	assert.equal(plan.totalCarbsG, sumCarbs);
	assert.equal(plan.totalFluidMl, sumFluid);
	assert.equal(plan.totalCarbsG, 120);
	assert.equal(plan.totalFluidMl, 1000);
});

test('kcal is estimated only when a bodyweight is supplied', () => {
	const without = buildFuelPlan(legs(), OPTS);
	assert.equal(without.legs[1].kcal, 0);
	const withWeight = buildFuelPlan(legs(), { ...OPTS, weightKg: 70 });
	// 1.036 kcal/kg/km × 70 kg × 8 km = 580.16
	assert.ok(Math.abs(withWeight.legs[1].kcal - 580.16) < 0.01);
});

test('non-positive intake rates clamp to zero', () => {
	const plan = buildFuelPlan(legs(), { carbsPerHourG: -10, fluidPerHourMl: -5 });
	assert.equal(plan.totalCarbsG, 0);
	assert.equal(plan.totalFluidMl, 0);
});
